p2p/peer.py

```diff
@@ -143,6 +143,106 @@
+        if block_number is None and not self.is_numbered:
+            raise TypeError(
+                "A `block_number` must be supplied to generate block numbers "
+                "for hash based header requests"
+            )
+        elif block_number is not None and self.is_numbered:
+            raise TypeError(
+                "The `block_number` parameter may not be used for number based "
+                "header requests"
+            )
+        elif block_number is None:
+            block_number = cast(BlockNumber, self.block_number_or_hash)
+
+        max_headers = min(eth.MAX_HEADERS_FETCH, self.max_headers)
+
+        # inline import until this module is moved to `trinity`
+        from trinity.utils.headers import sequence_builder
+        return sequence_builder(
+            block_number,
+            max_headers,
+            self.skip,
+            self.reverse,
+        )
+
+    @property
+    def is_numbered(self) -> bool:
+        return isinstance(self.block_number_or_hash, int)
+
+    def validate_headers(self,
+                         headers: Tuple[BlockHeader, ...]) -> None:
+        if not headers:
+            # empty response is always valid (TODO: deal with this)
+            return
+        elif not self.is_numbered:
+            first_header = headers[0]
+            if first_header.hash != self.block_number_or_hash:
+                raise ValueError("TODO: mismatched first block hash")
+
+        block_numbers: Tuple[BlockNumber, ...] = tuple(
+            header.block_number for header in headers
+        )
+        return self.validate_sequence(block_numbers)
+
+    def validate_sequence(self, block_numbers: Tuple[BlockNumber, ...]) -> None:
+        if not block_numbers:
+            return
+        elif self.is_numbered:
+            expected_numbers = self.generate_block_numbers()
+        else:
+            expected_numbers = self.generate_block_numbers(block_numbers[0])
+
+        # check for numbers that should not be present.
+        unexpected_numbers = set(block_numbers).difference(expected_numbers)
+        if unexpected_numbers:
+            raise ValueError(
+                'Unexpected numbers: {0}'.format(unexpected_numbers))
+
+        # check that the numbers are correctly ordered.
+        expected_order = tuple(sorted(
+            block_numbers,
+            reverse=self.reverse,
+        ))
+        if block_numbers != expected_order:
+            raise ValueError('Invalid ordering')
+
+        # check that all provided numbers are an ordered subset of the master
+        # sequence.
+        iter_expected = iter(expected_numbers)
+        for number in block_numbers:
+            for value in iter_expected:
+                if value == number:
+                    break
+            else:
+                raise ValueError('Unmatched number')
+
+    def is_valid_headers(self, headers: Tuple[BlockHeader, ...]) -> bool:
```

Comment:
This doesn't seem to be used anywhere

Reply:
I think we can drop this API with the planned change in the next iteration to only support a single in-flight request for a given command at any time.